torchacc/utils/optim_utils.py

```diff
@@ -0,0 +1,187 @@
+    if flatten_info != {}:
+        layer_name_list_ = []
+        layer_size_list_ = []
+        layer_numel_list_ = []
+        for name in layer_name_list:
+            if "_fsdp_wrapped_module.flat_param_" in name:
+                metadata = flatten_info[name]
+                prefix = ".".join(name.split(".")[:-1])
+                param_names, param_shapes, param_numel = metadata
+                full_names = param_names
+
+                if prefix != "":
+                    full_names = [prefix + "." + n for n in full_names]
+
+                full_names = [
+                    fn.replace("_fsdp_wrapped_module.",
+                               "").replace("_fpw_module.", "")
+                    for fn in full_names
+                ]
+
+                layer_name_list_.append(full_names)
+                layer_size_list_.append(param_shapes)
+                layer_numel_list_.append(param_numel)
+
+        return (layer_name_list_, layer_size_list_, layer_numel_list_)
+
+    # return with lists
+    layer_name_list = [[
+        fn.replace("_fsdp_wrapped_module.", "").replace("_fpw_module.", "")
+    ] for fn in layer_name_list]
+    layer_size_list = [[s] for s in layer_size_list]
+    layer_numel_list = [[n] for n in layer_numel_list]
+
+    return (layer_name_list, layer_size_list, layer_numel_list)
+
+
+def unflatten_optim_params(params, param_names, param_shapes, param_numels):
+    if params.dim() == 0:
+        full_params = [params for _ in range(len(param_names))]
+    else:
+        full_params = [
+            t.view(s)
+            for (t, s) in zip(params.split(param_numels), param_shapes)
+        ]
+
+    return full_params
+
+
+class _PosDimTensorInfo(NamedTuple):
+    """
+    Attributes:
+        shape (torch.Size): Sharded tensor shape (which is equal to the
+            unsharded tensor shape if the tensor is optimizer state for a
+            non-FSDP parameter and is hence not sharded).
+        dtype (torch.dtype): Data type of the tensor.
+    """
+
+    shape: torch.Size
+    dtype: torch.dtype
+
+
+def _setup_gloo_distributed(group):
+    pg = dist.new_group(ranks=group, backend="gloo")
+    return pg
+
+
+def _cleanup_gloo_distributed(pg):
+    dist.destroy_process_group(pg)
+
+
+def broadcast_processed_state(optim_state: dict[str, any], rank,
+                              sharding_groups):
+    objects: list[Any] = [None]
+    if rank == 0:
+        objects[0] = tree_map_only(
+            torch.Tensor,
+            lambda v: v.cpu() if v.dim() == 0 else _PosDimTensorInfo(
+                v.shape, v.dtype),  # type: ignore[union-attr]
+            optim_state,
+        )
+
+    ordinal = xm.get_ordinal()
+    new_group = []
+
+    for group in sharding_groups:
+        if ordinal in group:
+            new_group = group
+            break
+
+    pg_group = _setup_gloo_distributed(new_group)
+    dist.broadcast_object_list(objects, src=new_group[0], group=pg_group)
```

Comment:
Use `get_global_rank` instead, as we cannot assume that the ranks within the group are already sorted.

Reply:
done